### PR TITLE
Contest System + Featured Artist

### PR DESCRIPTION
**Pull request by nekodex**
Let's just say that this was ~~kinda~~ very rushed in order to get the contest+voting system out ASAP... (which is my way of saying that there's plenty of dodgy things done in the name of implementation speed)

Featured artist system included as contest voting is dependant on parts of that work (track player, etc) and has improvements/fixes built upon it... (I was working on featured artist stuff before contest voting became a priority)

(please limit feedback to critical things, rest will be fixed over the next week)

TL;DR: sorry, don't do as I did ; _ ;

### CODE REVIEW

File: resources/assets/coffee/react/contest-page.coffee [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+propsFunction = ->
+  tracks: osu.parseJson('json-tracks')
+  options:
+    hideGenre: true
+    hideBPM: true
+    hideDL: true
+    showVote: true
+    smaller: true
+
+reactTurbolinks.register 'tracklist', Tracklist, propsFunction
```
Comment (nanaya) [inline resources/assets/coffee/react/contest-page.coffee]: Pretty sure registering two components with same class name [doesn't work](/ppy/osu-web/blob/master/resources/assets/coffee/_classes/react-turbolinks.coffee#L39) unless turbolinks is disabled.

File: resources/assets/less/bem/artist.less [hunk 2]
```diff
@@ -0,0 +1,76 @@
+.artist {
+  &__description {
+    background-color: #293336;
+    color: white;
+    font-size: 12px;
+    font-weight: 100;
```
Comment (nanaya) [inline resources/assets/less/bem/artist.less]: Unless I'm missing something, exo 2 with font weight 100 isn't currently loaded.

File: resources/assets/less/bem/osu-page-header.less [hunk 3]
```diff
@@ -45,6 +76,27 @@
...
+  &--artist {
+    background-size: cover;
+    background-position: center;
+    height: 200px;
+    justify-content: flex-start;
+
+    font-weight: 100;
+    color: white;
+
+    &__subtitle {
```
Comment (nanaya) [inline resources/assets/less/bem/osu-page-header.less]: this is actually one place where nesting is allowed (or at least in my mind): `.block--modifier .block__element`.

Comment (nanaya) [inline resources/assets/less/bem/osu-page-header.less]: apparently there's even an faq entry for this http://getbem.com/faq/#block-modifier-affects-elements

File: resources/assets/less/variables.less [hunk 4]
```diff
@@ -63,6 +63,9 @@
 @font-size--header-subtitle: 16px;
 @font-size--header-description: 12px;
 
+@font-size--new-header-title: 30px;
```
Comment (nanaya) [inline resources/assets/less/variables.less]: wtf does 'new header' even mean ಠ_ಠ

File: resources/assets/less/bem/page-contents.less [hunk 5]
```diff
@@ -29,6 +29,15 @@
     flex-direction: row;
   }
 
+  &--artist {
+    flex-direction: column-reverse;
+    background-color: #293336;
```
Comment (nanaya) [inline resources/assets/less/bem/page-contents.less]: why are many of the color not using current palette?

File: resources/assets/coffee/react/contest-page.coffee (hunk 1 above)
Comment (nekodex) [inline resources/assets/coffee/react/contest-page.coffee]: yeah I suspected that, but only the contest stuff will be going live initially so _should_ be ok initially

File: resources/assets/less/bem/osu-page-header.less (hunk 3 above)
Comment (nekodex) [inline resources/assets/less/bem/osu-page-header.less]: :psy:

File: resources/assets/coffee/react/_components/tracklist.coffee [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+  constructor: (props) ->
+    super props
+    @state =
+      waitingForResponse: false
+      currently_playing: null
+      tracks: @props.tracks
+      voteCount: _.filter(@props.tracks, _.iteratee({selected: true})).length
+      options:
+        hideBPM: @props.options.hideBPM ?= false
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/tracklist.coffee]: this will change the props as well. Perhaps you mean just `@props.options.hideBPM ? false`?

File: resources/assets/less/variables.less (hunk 4 above)
Comment (nekodex) [inline resources/assets/less/variables.less]: new(er) (design) header

File: resources/assets/less/bem/page-contents.less (hunk 5 above)
Comment (nekodex) [inline resources/assets/less/bem/page-contents.less]: see previous comment about new design(s)

File: resources/assets/coffee/react/_components/tracklist.coffee [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+      data:
+        entry_id: @props.track.id
+
+    $.ajax laroute.route("contest.vote", contest_id: 1), params
+
+    .done (response) =>
+      $.publish 'tracklist:vote:done', tracks: response.tracks
+
+    .fail osu.ajaxError
+    .always LoadingOverlay.hide
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/tracklist.coffee]: missing `LoadingOverlay.show()`?

File: app/Http/Controllers/ContestsController.php [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    private function prepareTracks($contest)
+    {
+        $votes = [];
+        $seed = time();
+        if (Auth::check()) {
+            $user = Auth::user();
+            $seed = Auth::user()->user_id;
+            $votes = $contest->fresh(['votes'])->votes->where('user_id', $user->user_id);
```
Comment (nanaya) [inline app/Http/Controllers/ContestsController.php]: is the reason you needed this here because the one in `vote` doesn't work?

\### flash news start
While `where()` and other stuff modifies current query builder, `fresh()` doesn't change the current object.

\### flash news end

Thus you probably want `$contents = $contest->fresh(['votes'])` for the one in `vote` function above.

**Comment by nanaya** [comment]
is it just me or this thing is easier without react?

File: resources/assets/coffee/react/_components/tracklist.coffee [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    super props
+    # @sendVote = _.debounce @sendVote, 500
+
+  sendVote: =>
+    params =
+      method: 'PUT'
+      data:
+        entry_id: @props.track.id
+
+    $.ajax laroute.route("contest.vote", contest_id: 1), params
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/tracklist.coffee]: cool contest id ｡：ﾟ(｡ﾉω＼｡)ﾟ･｡

**Comment by nekodex** [comment]
I did it initially without but thought using react would make the voting stuff easier...  ┐(´д｀)┌

File: resources/views/contests/show.blade.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+@section("script")
+  @parent
+
+  <script id="json-tracks" type="application/json">
+    {!! json_encode($tracks) !!}
+  </script>
+
+  <script id="json-contest" type="application/json">
+    {!! json_encode(['id' => $contest->id]) !!}
```
Comment (nanaya) [inline resources/views/contests/show.blade.php]: pass `$contest_id` to view, and then use `$contest->id` :+1:

File: app/Http/Controllers/ContestsController.php [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+    public function vote($id)
+    {
+        $user = Auth::user();
+        $contest = Contest::with('entries')->findOrFail($id);
+        $entry = $contest->entries()->findOrFail(Request::input('entry_id'));
+
+        priv_check('ContestVote', $contest)->ensureCan();
+
+        $contest->vote($user, $entry);
```
Comment (nanaya) [inline app/Http/Controllers/ContestsController.php]: personally I'd make adding/removing vote an explicit parameter.

**Comment by peppy** [comment]
merging on @nanaya's request
